cudax/include/cuda/experimental/__stf/utility/unittest.cuh

```diff
@@ -0,0 +1,648 @@
+    ::std::experimental::source_location::current(), ::cuda::experimental::stf::expecter()->*__VA_ARGS__)
+
+namespace cuda::experimental::stf
+{
+
+/*
+ * @brief Entry point for all comparisons (equality, inequality, and ordering) tested.
+ *
+ */
+struct expecter
+{
+  /*
+   * @brief Exception embedding location information.
+   *
+   */
+  struct failure : ::std::runtime_error
+  {
+    template <typename... Msgs>
+    failure(::std::experimental::source_location loc, const Msgs&... msgs)
+        : ::std::runtime_error(text(loc, msgs...))
+    {}
+
+    template <typename... Msgs>
+    static ::std::string text(::std::experimental::source_location loc, const Msgs&... msgs)
+    {
+      ::std::stringstream s;
+      s << loc.file_name() << '(' << loc.line() << "): ";
+      (stream(s, msgs), ...);
+      auto result = s.str();
+      if (!result.empty() && result.back() != '\n')
+      {
+        result += '\n';
+      }
+      return result;
+    }
+
+  private:
+    template <typename T>
+    static void stream(::std::stringstream& s, const T& x)
+    {
+      if constexpr (has_ostream_operator<T>::value)
+      {
+        s << x;
+      }
+      else
+      {
+        s << type_name<T> << "{?}";
+      }
+    }
+  };
+
+  /*
+   * @brief Wrapper for a comparison operation using one of `==`, `!=`, `<`, `>`, `<=`, `>=`. Includes the result and
+   * the operands.
+   *
+   * @tparam T Type of the left-hand side operator (qualifiers and reference included)
+   * @tparam U Type of the right-hand side operator (qualifiers and reference included)
+   */
+  template <typename L, typename R>
+  struct comparison_expression
+  {
+    bool value;
+    L lhs;
+    R rhs;
+    const char* op;
+    comparison_expression(bool value, L lhs, R rhs, const char* op)
+        : value(value)
+        , lhs(::std::forward<L>(lhs))
+        , rhs(::std::forward<R>(rhs))
+        , op(op)
+    {}
+  };
+
+  /*
+   * @brief Wraps a single term (the left-hand side one) in a comparison expression.
+   *
+   * @tparam T type of the term, may include a reference or a const reference etc.
+   */
+  template <typename T>
+  struct term
+  {
+    T value;
+
+    template <typename U>
+    term(U&& value)
+        : value(::std::forward<U>(value))
+    {}
+
+    /*
+     * All comparison operators are defined assuming `const&` inputs. They are evaluated eagerly and return a
+     * `compariso_expression` object with information necessary for constructing an error message, if any.
+     */
+#define _9d10c7e37932af3c4f39a5ce7ff00b5a(op)                                         \
+  template <typename U>                                                               \
+  auto operator op(const U& rhs)&&                                                    \
+  {                                                                                   \
+    const bool r = value op rhs;                                                      \
```

Comment:
This breaks compilation due to comparing different types now, we should revert that change and get it work